Looks like there is a quite unrelated failure in the kinetis frdm board build.

I don't think it should block.

Anyone have privs to merge anyway?
```
[117/138] Building C object zephyr/drivers/adc/CMakeFiles/drivers__adc.dir/adc_shell.c.obj
FAILED: zephyr/drivers/adc/CMakeFiles/drivers__adc.dir/adc_shell.c.obj 
ccache /opt/sdk/zephyr-sdk-0.11.2/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc -DBOARD_FLASH_SIZE=CONFIG_FLASH_SIZE -DBUILD_VERSION=zephyr-v2.2.0-2427-gbf9c4ea1568e -DCPU_MK64FN1M0VLL12 -DKERNEL -DNDEBUG -D_FORTIFY_SOURCE=2 -D__PROGRAM_START -D__ZEPHYR_SUPERVISOR__ -D__ZEPHYR__=1 -I../../../../../../include -Izephyr/include/generated -I../../../../../../soc/arm/nxp_kinetis/k6x -I../../../../../../lib/fnmatch/. -I/home/buildslave/src/github.com/zephyrproject-rtos/modules/hal/cmsis/CMSIS/Core/Include -I/home/buildslave/src/github.com/zephyrproject-rtos/modules/hal/nxp/mcux/devices/MK64F12 -I/home/buildslave/src/github.com/zephyrproject-rtos/modules/hal/nxp/mcux/drivers/kinetis/. -isystem ../../../../../../lib/libc/minimal/include -isystem /opt/sdk/zephyr-sdk-0.11.2/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/9.2.0/include -isystem /opt/sdk/zephyr-sdk-0.11.2/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/9.2.0/include-fixed -Os -imacros/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/sanity-out/frdm_k64f/tests/boards/board_shell/boards.shell.shell/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -mcpu=cortex-m4 -mthumb -mabi=aapcs -imacros/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/include/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-address-of-packed-member -Wno-pointer-sign -Wpointer-arith -Wno-unused-but-set-variable -Werror=implicit-int -Werror -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-strict-overflow -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/tests/boards/board_shell=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/buildslave/src/github.com/zephyrproject-rtos/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/home/buildslave/src/github.com/zephyrproject-rtos=WEST_TOPDIR -ffunction-sections -fdata-sections -std=c99 -nostdinc -MD -MT zephyr/drivers/adc/CMakeFiles/drivers__adc.dir/adc_shell.c.obj -MF zephyr/drivers/adc/CMakeFiles/drivers__adc.dir/adc_shell.c.obj.d -o zephyr/drivers/adc/CMakeFiles/drivers__adc.dir/adc_shell.c.obj   -c /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/adc/adc_shell.c
In file included from /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/adc/adc_shell.c:7:
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/adc/adc_shell.c:477:22: error: expected expression before ',' token
  477 |  ADC_SHELL_COMMAND(0),
      |                      ^
../../../../../../include/shell/shell.h:219:3: note: in definition of macro 'SHELL_STATIC_SUBCMD_SET_CREATE'
  219 |   __VA_ARGS__      \
      |   ^~~~~~~~~~~
```

--- a/tests/kernel/pipe/pipe_api/src/main.c
+++ b/tests/kernel/pipe/pipe_api/src/main.c
@@ -32,6 +32,12 @@ extern void test_pipe_user_get_fail(void);
 extern void test_resource_pool_auto_free(void);
 #endif
 
+extern void test_pipe_avail_r_lt_w(void);
+extern void test_pipe_avail_w_lt_r(void);
+extern void test_pipe_avail_r_eq_w_full(void);
+extern void test_pipe_avail_r_eq_w_empty(void);
+extern void test_pipe_avail_no_buffer(void);
+
 /* k objects */
 extern struct k_pipe pipe, kpipe, khalfpipe, put_get_pipe;
 extern struct k_sem end_sema;
@@ -77,6 +83,11 @@ void test_main(void)
 			 ztest_unit_test(test_half_pipe_saturating_block_put),
 			 ztest_1cpu_unit_test(test_pipe_alloc),
 			 ztest_unit_test(test_pipe_reader_wait),
-			 ztest_1cpu_unit_test(test_pipe_block_writer_wait));
+			 ztest_1cpu_unit_test(test_pipe_block_writer_wait),
+			 ztest_unit_test(test_pipe_avail_r_lt_w),
+			 ztest_unit_test(test_pipe_avail_w_lt_r),
+			 ztest_unit_test(test_pipe_avail_r_eq_w_full),
+			 ztest_unit_test(test_pipe_avail_r_eq_w_empty),
+			 ztest_unit_test(test_pipe_avail_no_buffer));
 	ztest_run_test_suite(pipe_api);
 }

--- a/tests/kernel/pipe/pipe_api/src/test_pipe_avail.c
+++ b/tests/kernel/pipe/pipe_api/src/test_pipe_avail.c
@@ -1,0 +1,215 @@
+/*
+ * Copyright (c) 2020 Friedt Professional Engineering Services, Inc
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @brief Tests for the Pipe read / write availability
+ * @ingroup kernel_pipe_tests
+ * @{
+ */
+
+#include <ztest.h>
+
+static ZTEST_DMEM unsigned char __aligned(4) data[] = "abcdefgh";
+static struct k_pipe pipe = {
+	.buffer = data,
+	.size = sizeof(data) - 1 /* '\0' */,
+};
+
+static struct k_pipe bufferless;
+
+/**
+ * @brief Ensure that bufferless pipes return 0 bytes available
+ *
+ * Pipes can be created to be bufferless (i.e. @ref k_pipe.buffer is `NULL`
+ * and @ref k_pipe.size is 0).
+ *
+ * If either of those conditions is true, then @ref k_pipe_read_avail and
+ * @ref k_pipe_write_avail should return 0.
+ *
+ * @note
+ * A distinction can be made between buffered and bufferless pipes in that
+ * @ref k_pipe_read_avail and @ref k_pipe_write_avail will never
+ * simultaneously return 0 for a buffered pipe, but they will both return 0
+ * for an unbuffered pipe.
+ */
+void test_pipe_avail_no_buffer(void)
+{
+	size_t r_avail;
+	size_t w_avail;
+
+	r_avail = k_pipe_read_avail(&bufferless);
+	zassert_equal(r_avail, 0, "read: expected: 0 actual: %u", r_avail);
+
+	w_avail = k_pipe_write_avail(&bufferless);
+	zassert_equal(w_avail, 0, "write: expected: 0 actual: %u", w_avail);
+}
+
+/**
+ * @brief Test available read / write space for r < w
+ *
+ * This test case is for buffered @ref k_pipe objects and covers the case
+ * where @ref k_pipe.read_index is less than @ref k_pipe.write_index.
+ *
+ * In this case, @ref k_pipe.bytes_used is not relevant.
+ *
+ *      r     w
+ *     |a|b|c|d|e|f|g|h|
+ *     |0|1|2|3|4|5|6|7|
+ *
+ * As shown above, the pipe will be able to read 3 bytes without blocking
+ * and write 5 bytes without blocking.
+ *
+ * Thus
+ *     r_avail = w - r = 3
+ *     would read: a b c d
+ *
+ *     w_avail = N - (w - r) = 5
+ *     would overwrite: e f g h
+ */
+void test_pipe_avail_r_lt_w(void)
+{
+	size_t r_avail;
+	size_t w_avail;
+
+	pipe.read_index = 0;
+	pipe.write_index = 3;
+	/* pipe.bytes_used is irrelevant */
+
+	r_avail = k_pipe_read_avail(&pipe);
+	zassert_equal(r_avail, 3, "read: expected: 3 actual: %u", r_avail);
+
+	w_avail = k_pipe_write_avail(&pipe);
+	zassert_equal(w_avail, 5, "write: expected: 5 actual: %u", w_avail);
+}
+
+/**
+ * @brief Test available read / write space for w < r
+ *
+ * This test case is for buffered @ref k_pipe objects and covers the case
+ * where @ref k_pipe.write_index is less than @ref k_pipe.read_index.
+ *
+ * In this case, @ref k_pipe.bytes_used is not relevant.
+ *
+ *      w     r
+ *     |a|b|c|d|e|f|g|h|
+ *     |0|1|2|3|4|5|6|7|
+ *
+ *
+ * As shown above, the pipe will fbe able to read 5 bytes without blocking
+ * and write 3 bytes without blocking.
+ *
+ * Thus
+ *     r_avail = N - (r - w) = 5
+ *     would read: e f g h
+ *
+ *     w_avail = r - w = 3
+ *     would overwrite: a b c d
+ */
+void test_pipe_avail_w_lt_r(void)
+{
+	size_t r_avail;
+	size_t w_avail;
+
+	pipe.read_index = 3;
+	pipe.write_index = 0;
+	/* pipe.bytes_used is irrelevant */
+
+	r_avail = k_pipe_read_avail(&pipe);
+	zassert_equal(r_avail, 5, "read: expected: 4 actual: %u", r_avail);
+
+	w_avail = k_pipe_write_avail(&pipe);
+	zassert_equal(w_avail, 3, "write: expected: 4 actual: %u", w_avail);
+}
+
+/**
+ * @brief Test available read / write space for `r == w` and an empty buffer
+ *
+ * This test case is for buffered @ref k_pipe objects and covers the case
+ * where @ref k_pipe.read_index is equal to @ref k_pipe.write_index and
+ * @ref k_pipe.bytes_used is zero.
+ *
+ * In this case, @ref k_pipe.bytes_used is relevant because the read and
+ * write indeces are equal.
+ *
+ *            r
+ *            w
+ *     |a|b|c|d|e|f|g|h|
+ *     |0|1|2|3|4|5|6|7|
+ *
+ * Regardless of whether the buffer is full or empty, the following holds:
+ *
+ *     r_avail = bytes_used
+ *     w_avail = N - bytes_used
+ *
+ * Thus:
+ *     r_avail = 0
+ *     would read:
+ *
+ *     w_avail = N - 0 = 8
+ *     would overwrite: e f g h a b c d
+ */
+void test_pipe_avail_r_eq_w_empty(void)
+{
+	size_t r_avail;
+	size_t w_avail;
+
+	pipe.read_index = 4;
+	pipe.write_index = 4;
+	pipe.bytes_used = 0;
+
+	r_avail = k_pipe_read_avail(&pipe);
+	zassert_equal(r_avail, 0, "read: expected: 0 actual: %u", r_avail);
+
+	w_avail = k_pipe_write_avail(&pipe);
+	zassert_equal(w_avail, 8, "write: expected: 8 actual: %u", w_avail);
+}
+
+/**
+ * @brief Test available read / write space for `r == w` and a full buffer
+ *
+ * This test case is for buffered @ref k_pipe objects and covers the case
+ * where @ref k_pipe.read_index is equal to @ref k_pipe.write_index and
+ * @ref k_pipe.bytes_used is equal to @ref k_pipe.size.
+ *
+ * In this case, @ref k_pipe.bytes_used is relevant because the read and
+ * write indeces are equal.
+ *
+ *            r
+ *            w
+ *     |a|b|c|d|e|f|g|h|
+ *     |0|1|2|3|4|5|6|7|
+ *
+ * Regardless of whether the buffer is full or empty, the following holds:
+ *
+ *     r_avail = bytes_used
+ *     w_avail = N - bytes_used
+ *
+ * Thus
+ *     r_avail = N = 8
+ *     would read: e f g h a b c d
+ *
+ *     w_avail = N - 8 = 0
+ *     would overwrite:
+ */
+void test_pipe_avail_r_eq_w_full(void)
+{
+	size_t r_avail;
+	size_t w_avail;
+
+	pipe.read_index = 4;
+	pipe.write_index = 4;
+	pipe.bytes_used = pipe.size;
+
+	r_avail = k_pipe_read_avail(&pipe);
+	zassert_equal(r_avail, 8, "read: expected: 8 actual: %u", r_avail);
+
+	w_avail = k_pipe_write_avail(&pipe);
+	zassert_equal(w_avail, 0, "write: expected: 0 actual: %u", w_avail);
+}
+
+/**
+ * @}
+ */

